### PR TITLE
Eliminate multiple client creations

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn/longhorn#6866
Longhorn/longhorn#6936

### CODE REVIEW

**Comment by derekbit** [comment]
The memory usage is reduced by 50% (800MiB->400MiB) in the 1000 RWO volumes test.

**Review by innobead — APPROVED** [review]
LGTM.

File: util/client/client.go [hunk 1]
```diff
@@ -1,75 +1,111 @@
...
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get clientset")
+	}
 
-		kubeInformerFactory := informers.NewSharedInformerFactory(kubeClient, time.Second*30)
-		lhInformerFactory := lhinformers.NewSharedInformerFactory(lhClient, time.Second*30)
+	scheme := runtime.NewScheme()
+	if err := longhorn.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return nil, errors.Wrap(err, "unable to create scheme")
```
Comment (derekbit) [inline util/client/client.go:71]: One question
Is it ok that call `AddToScheme` twice here?
cc @innobead @shuo-wu

File: webhook/server/server.go [hunk 2]
```diff
@@ -47,31 +45,24 @@ var (
 
 type WebhookServer struct {
 	context     context.Context
-	cfg         *rest.Config
 	namespace   string
 	webhookType string
 }
 
-func New(ctx context.Context, cfg *rest.Config, namespace, webhookType string) *WebhookServer {
+func New(ctx context.Context, namespace, webhookType string) *WebhookServer {
```
Comment (shuo-wu) [inline webhook/server/server.go:52]: NIT: Why not cache `clients` in this `WebhookServer`? Then the below methods do not need input `clients *client.Clients`

**Review by shuo-wu — APPROVED** [review]
LGTM

File: webhook/server/server.go (hunk 2 above)
Comment (derekbit) [reply webhook/server/server.go]: Updated. Thank you.

File: util/client/client.go (hunk 1 above)
Comment (ChanYiLin) [reply util/client/client.go:71]: I think is okay to call it twice
AddToScheme function should be idempotent

File: util/client/client.go [hunk 3]
```diff
@@ -1,75 +1,111 @@
...
+	wranglerSchemes "github.com/rancher/wrangler/pkg/schemes"
+	"github.com/sirupsen/logrus"
 
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/informers"
-	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
 
-	corev1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
```
Comment (innobead) [inline util/client/client.go:17]: Just noticed the previous is `corev1 "k8s.io/api/apps/v1"`, so is there any typo before?

Comment (innobead) [reply util/client/client.go:17]: cc @c3y1huang is this caused by a wrong change by a previous format update?

Comment (ChanYiLin) [reply util/client/client.go:17]: I think both 
"k8s.io/api/core/v1", "k8s.io/api/apps/v1" have `AddToScheme` function and they are actually pointing to the same place.
So both works fine, but in this PR, we need to use `corev1.NamespaceDefault` which  only in "k8s.io/api/core/v1" library

**Review by ChanYiLin — APPROVED** [review]
LGTM

File: webhook/server/server.go [hunk 4]
```diff
@@ -121,9 +109,9 @@ func (s *WebhookServer) ListenAndServe() error {
 	}
 }
 
-func (s *WebhookServer) runAdmissionWebhookListenAndServe(client *client.Client, handler http.Handler, validationResources []admission.Resource, mutationResources []admission.Resource) error {
-	apply := client.Apply.WithDynamicLookup()
-	client.Core.Secret().OnChange(s.context, "secrets", func(key string, secret *corev1.Secret) (*corev1.Secret, error) {
+func (s *WebhookServer) runAdmissionWebhookListenAndServe(clients *client.Clients, handler http.Handler, validationResources []admission.Resource, mutationResources []admission.Resource) error {
```
Comment (shuo-wu) [inline webhook/server/server.go:112]: NIT: No need to input `clients *client.Clients`. Same below

File: util/client/client.go (hunk 3 above)
Comment (innobead) [reply util/client/client.go:17]: It's different. One is for `app` workload related resource scheme but `core` is for basic resources like the pod below. 

[/vendor/k8s.io/api/core/v1/register.go#L36-L49](https://github.com/longhorn/longhorn-manager/blob/a6a15ae9f4546f796bd4ebdc31851536d11a8c8b/vendor/k8s.io/api/core/v1/register.go#L36-L49)
```go
var (
	// We only register manually written functions here. The registration of the
	// generated functions takes place in the generated files. The separation
	// makes the code compile even when the generated files are missing.
	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
	AddToScheme   = SchemeBuilder.AddToScheme
)

// Adds the list of known types to the given scheme.
func addKnownTypes(scheme *runtime.Scheme) error {
	scheme.AddKnownTypes(SchemeGroupVersion,
		&Pod{},
		&PodList{},
		&PodStatusResult{},
```

Comment (c3y1huang) [reply util/client/client.go:17]: > cc @c3y1huang is this caused by a wrong change by a previous format update?

It was previously [v1 k8s.io/api/apps/v1](https://github.com/longhorn/longhorn-manager/commit/c93aea9de7ed5e02feafcd70d3888d75742571ba#diff-7b3f4aabf038dc0bcb882dac26677c633b8cbb05ee4b57cfcbff61ea5051ec56L12).

Comment (innobead) [reply util/client/client.go:17]: I see. Let's fix this if it has been backported.

Comment (c3y1huang) [reply util/client/client.go:17]: https://github.com/longhorn/longhorn-manager/pull/2238

Comment (innobead) [reply util/client/client.go:17]: It's confusing, that we were using is core resource (secret below) but used the app scheme.

[/webhook/server/server.go#L126-L131](https://github.com/longhorn/longhorn-manager/blob/a6a15ae9f4546f796bd4ebdc31851536d11a8c8b/webhook/server/server.go#L126-L131)
```go
	client.Core.Secret().OnChange(s.context, "secrets", func(key string, secret *corev1.Secret) (*corev1.Secret, error) {
		if secret == nil || secret.Name != caName || secret.Namespace != s.namespace || len(secret.Data[corev1.TLSCertKey]) == 0 {
			return nil, nil
		}

		port := int32(types.DefaultWebhookServerPort)

```

File: webhook/server/server.go (hunk 4 above)
Comment (derekbit) [reply webhook/server/server.go]: Updated. Thank you.

File: util/client/client.go (hunk 3 above)
Comment (derekbit) [reply util/client/client.go:18]: Fixed. Thank you.

**Review by innobead — APPROVED** [review]
LGTM.

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Pull request merged**